feat: better error message when barcode length differs from expected

TimD1 opened the pull request:
- fixes #31, error message is now more clear

nh13 commented on src/lib/barcode_matching.rs:
```diff
@@ -74,7 +74,7 @@ impl BarcodeMatcher {
             observed_bases.len(),
             expected_bases.len(),
-            "observed_bases: {}, expected_bases: {}",
+            "Sample barcode length ({}) differs from expected ({}) during barcode matching",
```
I think we want to know which _sample_ this is caused by, as well as the barcode bases (actual and expected) along with the length.

TimD1 replied:
Sounds good. I can add that info if I replace `BarcodeMatcher.sample_barcodes` with `BarcodeMatcher.samples` (vector of `Sample` structs storing `sample_id`s and `barcode`s). It looks like this was the eventual plan anyways: https://github.com/fulcrumgenomics/fqtk/blob/33925525a79dfb8cc1ca23ba78e6c2f0e59eda42/src/lib/barcode_matching.rs#L25-L27

TimD1 commented:
This is my first time writing Rust, so please feel free to make suggestions for better style/design/etc.

A note on the new tests:
Since `Sample`s do not allow empty barcodes:
https://github.com/fulcrumgenomics/fqtk/blob/6ed98a3334eb13f22dd8054d39d9b8f77e76b337/src/lib/samples.rs#L48-L50

My understanding is that `new()` is the only way to get a new `Sample` outside of the `samples.rs` module, since `Sample` contains a private field (`ordinal`). Therefore, I dropped the test for `BarcodeMatcher` being initialized with a `Sample` with an empty barcode:
https://github.com/fulcrumgenomics/fqtk/blob/6ed98a3334eb13f22dd8054d39d9b8f77e76b337/src/lib/barcode_matching.rs#L178-L185

And modified the test that allowed empty barcodes to be compared to check that it now fails.
https://github.com/fulcrumgenomics/fqtk/blob/6ed98a3334eb13f22dd8054d39d9b8f77e76b337/src/lib/barcode_matching.rs#L195-L198

nh13 commented on src/lib/barcode_matching.rs:
```diff
@@ -37,31 +35,31 @@ pub struct BarcodeMatcher {
-            .iter()
-            .map(|barcode| BString::from(barcode.to_ascii_uppercase()))
-            .collect::<Vec<_>>();
+        let mut modified_samples = samples.to_vec();
+        for sample in modified_samples.iter_mut() {
+            sample.barcode = sample.barcode.to_ascii_uppercase();
+        }
```
Shall we make a clone of the sample instead of modifying them?

TimD1 replied:
I think `to_vec()` should already create a copy of `samples`: https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec

I just rebased and pushed my changes.

nh13 reviewed (changes requested):
One minor question

nh13 reviewed (approved):
Thanks @TimD1 !
